kubessh/sftp.py

```diff
@@ -0,0 +1,164 @@
+        self._proc_stdin.write(data)
+        self._proc_stdin.flush()
+
+    async def _read_packet_from_delegate(self):
+        self.logger.debug1('read packet from delegate process')
+        ch = self._proc_stdout
+        self.logger.debug2('read packet length field')
+        pkt_len = int.from_bytes(ch.peek()[:4], byteorder='big', signed=False)
+        self.logger.debug2(f'attempt to read {pkt_len} byte packet')
+        packet = ch.read(pkt_len + 4)
+        self.logger.debug2(f'delegate packet is {packet}')
+        return packet
+
+    async def _forward_outgoing_packet_from_delegate(self, data):
+        self._writer.write(data)
+
+    async def _process_packet(self, pkttype, pktid, packet):
+        self.log_received_packet(pkttype, pktid, packet, note='forwarded to subprocess')
+        await self._forward_incoming_packet_to_delegate(packet.get_full_payload())
+        response = await self._read_packet_from_delegate()
+        # TODO: (would have to redundantly decode packet to log it, omitted for now)
+        #self.log_sent_packet(pkttype, pktid, packet, note='forwarded from subprocess')
+        await self._forward_outgoing_packet_from_delegate(response)
+
+    async def run(self):
+        # skip super's SFTP init handling, the subprocess does this here
+        await self.recv_packets()
+
+
+class KubeSFTPServerHandler(DelegateSFTPServerHandler, LoggingConfigurable):
+
+    # where the SFTP binary lives in the KubeSSH server image
+    sftp_binary_source_path = Unicode(
+        '/usr/local/share/kubessh/sftp-server',
+        allow_none=False,
+        help="""
+        Path in the KubeSSH server container to the static sftp-server binary.
+        """,
+        config=True
+    )
+
+    # where the SFTP binary goes in the user pod
+    sftp_binary_upload_path = Unicode(
+        '.local/sbin/kubessh-sftp-server',
+        allow_none=False,
+        help="""
+        Path inside the user pod where the sftp-server binary will be uploaded.
+
+        Must be full path including the executable name, and can be absolute or
+        relative to the working dir. Must be writable by the user the pod runs
+        as.
+        """,
+        config=True
+    )
+
+    def __init__(self, server, reader, writer):
+        from kubessh.app import app
+        super(LoggingConfigurable, self).__init__(parent=app)
+        super().__init__(server, reader, writer)
+
+    def _run_setup_command(self, *kubectl_command):
+        self.logger.info(f'executing: {kubectl_command}')
+        cmd = subprocess.Popen(kubectl_command)
+        return cmd.wait()
+
+    def _init_delegate(self):
+        self.namespace = self.parent.default_namespace
+        self.pod_name = UserPod(
+            parent=self.parent, namespace=self.namespace,
+            username=self.username).pod_name
+        # ensure sftp-server binary is copied into the user pod
+        self._run_setup_command(
```

Comment:
Instead of running kubectl cp, how about:

1. We create a small container image that *just* has sftp-binary
2. Set up an emptyDir volume by default in our launched pod
3. Run an initContainer with our sftp-binary image, where we copy the binary to the emptyDir.
4. When running kubectl exec, we just refer to the binary path in the emptyDir volume, which we will know.

This has a few advantages over kubectl cp:

1. sftp-binary image is pulled only once per node, and should be pretty small
2. The initContainer should execute pretty quickly, since the image is already there
3. We can put other binaries (or whole packages) we want in the emptyDir volume later if we need
4. SFTP can begin immediately, without us needing to copy a binary in.

What do you think of this suggestion?

Reply:
Great idea! Plus, it removes the need to have `tar` available in the user image, as would be needed for `kubectl cp` to work.